controlpanel/frontend/static/components/selectable-rows/selectable-rows.js

```diff
@@ -1,24 +1,92 @@
+/**
+ * Module to allow selecting entire rows in a table instead of using a small checkbox.
+ *
+ * It will disable any in-row actions buttons once a selection is made, but will re-enable them once
+ * all the selections are cleared.
+ *
+ * It can optionally support enabling a (global)
+ * button(s) that lives outside the table.
+ * How to use:
+ *  1. Add the class "selectable-rows" to your <table>
+ *  2. For any button you want to enable when there are selections add the class "selectable-rows__enable-on-selections"
+ *  3. Use the macro `row_selector` to put insert an invisible input box in the first column of your table
+ * The module will automatically disable any action buttons that are in side the row.
+ *
+ * WARNING: This module doesn't support more than 1 table on the same page
+ * @type {{init(): void, onSelectionChanged(*): void, baseClass: string, bindEvents(): void}}
+ */
+
 moj.Modules.selectable_rows = {
-  selector: '.selectable-rows tr',
+  baseClass: "selectable-rows",
 
   init() {
-    if (document.querySelector(this.selector)) {
+    this.tableRowSelector = `.${this.baseClass} tr`;
+    this.tableRowButtonSelector = `${this.tableRowSelector} button`;
+    this.selectionClass = `${this.baseClass}__row-selected`;
+    this.enableOnSelectionSelector = `.${this.baseClass}__enable-on-selections`;
+    this.selectionChangedEventName = `${this.baseClass}:rowSelectionChanged`;
+    if (document.querySelector(this.tableRowSelector)) {
       this.bindEvents();
     }
   },
 
+  /**
+   * Runs when rows are selected/deselected
+   * @param event
+   */
+  onSelectionChanged(event) {
+    const rowButtons = document.querySelectorAll(this.tableRowButtonSelector);
+    const enableOnSelectionButtons = document.querySelectorAll(
+      this.enableOnSelectionSelector
+    );
+    if (event.detail.rowsSelected.length > 0) {
+      // some rows are selected
+      rowButtons.forEach(e => (e.disabled = true));
+      enableOnSelectionButtons.forEach(e => e.disabled = false);
+    } else {
+      rowButtons.forEach(e => (e.disabled = false));
+      enableOnSelectionButtons.forEach(e => e.disabled = true);
+    }
+  },
+
   bindEvents() {
-    document.querySelectorAll(this.selector).forEach(row => {
+    window.addEventListener(this.selectionChangedEventName, e =>
```

Comment:
I may miss something but couldn't just pass `this.onSelectionChanged` as event listener instead of the anonymous function which calls it? the event loop would call it anyway passing the event right?

Reply:
I think I had it like that to begin with but then I think `this` was bound to something else so using the => arrow syntax to keep `this` the same

Reply:
interesting, that is so sneaky!

Reply:
more info: https://maximdenisov.gitbooks.io/you-don-t-know-js/content/scope_&_closures/lexical-this.html

Reply:
I doubt this will make the cut in a newer edition of "Javascript: The Good Parts".

Reply:
I've encountered this before but I still find it sneaky. "arrow functions as pretty much like normal functions...but WAIT, not quite".

In a way the weirdness factor is caused by the fact the behaviour of the old school functions is weird and the new ones are behaving in a better, more predictable way.

Oh well, it's JS 🤷‍♂